src/libraries/System.Private.CoreLib/src/System/Reflection/InvokeSignatureInfo.cs

```diff
@@ -0,0 +1,207 @@
+                !ReferenceEquals(_returnType, otherSig._returnType) ||
+                _isStatic != otherSig._isStatic ||
+                _parameterTypes.Length != otherSig._parameterTypes.Length)
+            {
+                return false;
+            }
+
+            for (int i = 0; i < _parameterTypes.Length; i++)
+            {
+                if (!ReferenceEquals(_parameterTypes[i], otherSig._parameterTypes[i]))
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        public override int GetHashCode() => GetHashCode(_declaringType, _parameterTypes, _returnType);
+
+        public static int GetHashCode(Type? declaringType, Type[] parameterTypes, Type returnType)
+        {
+            int hashcode = 0;
+            if (declaringType is not null)
+            {
+                hashcode = int.RotateLeft(declaringType.GetHashCode(), 5);
+            }
+
+            hashcode = int.RotateLeft(hashcode ^ returnType.GetHashCode(), 5);
+
+            for (int i = 0; i < parameterTypes.Length; i++)
+            {
+                hashcode = int.RotateLeft(hashcode ^ parameterTypes[i].GetHashCode(), 5);
+            }
+
+            // We don't include _isStatic because for normalized methods it is already included with _declaringType==typeof(void)
+            // and for unnormalized methods it is not possible to have an instance and static method with the same name.
+            return hashcode;
+        }
+    }
+
+    /// <summary>
+    /// Provide a zero-alloc ref struct to wrap member signature properties needed for both cache lookup and emit.
+    /// </summary>
+    internal readonly ref struct InvokeSignatureInfoKey
+    {
+        internal readonly Type? _declaringType;
+        internal readonly Type[] _parameterTypes;
+        internal readonly Type _returnType;
+        internal readonly bool _isStatic;
+
+        public static InvokeSignatureInfoKey CreateNormalized(Type declaringType, Type[] parameterTypes, Type returnType, bool isStatic)
+        {
+            return new InvokeSignatureInfoKey(
+                isStatic ? typeof(void) : MakeNormalized(declaringType),
+                GetNormalizedParameterTypes(parameterTypes),
+                MakeNormalized(returnType),
+                isStatic);
+        }
+
+        public InvokeSignatureInfoKey(Type? declaringType, Type[] parameterTypes, Type returnType, bool isStatic)
+        {
+            _declaringType = declaringType;
+            _parameterTypes = parameterTypes;
+            _returnType = returnType;
+            _isStatic = isStatic;
+        }
+
+        public Type? DeclaringType => _declaringType;
+        public Type[] ParameterTypes => _parameterTypes;
+        public Type ReturnType => _returnType;
+        public bool IsStatic => _isStatic;
+
+        public static bool AlternativeEquals(in InvokeSignatureInfoKey @this, InvokeSignatureInfo signatureInfo)
```

Comment:
We should avoid the `@this` pattern. It is rarely used in the BCL and I personally dislike it. The english language has enough words that we can avoid using keywords. In this case, I toulw suggest `inst`, `instance`, or `key`.

Reply:
Sure thanks.